python/perf-kernels/rmsnorm.py

```diff
@@ -37,100 +46,96 @@ def get_autotune_config():
-        cols = start + col_offsets
-        input_ptrs = row_input_ptr + cols
-        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
-        x = tl.load(input_ptrs)
-        sum_squares += tl.sum(x * x, axis=0)
-
-    # loop peeling for mask
-    cols = n_cols_blks * BLOCK_SIZE + col_offsets
-    mask = cols < n_cols
-    input_ptrs = row_input_ptr + cols
-    input_ptrs = tl.multiple_of(input_ptrs, (16, ))
-    x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
-    sum_squares += tl.sum(x * x, axis=0)
-
-    return sum_squares
-
-
-# apply normalization to each block of the row
-@triton.jit
-def apply_normalization(input_ptr, output_ptr, g_ptr, input_row_stride, output_row_stride, n_cols, norm_factor,
-                        BLOCK_SIZE, row_idx):
-    col_offsets = tl.arange(0, BLOCK_SIZE)
-    row_input_ptr = input_ptr + row_idx * input_row_stride
-    row_output_ptr = output_ptr + row_idx * output_row_stride
-
-    for start in range(0, n_cols, BLOCK_SIZE):
-        cols = start + col_offsets
-        mask = cols < n_cols
-        input_ptrs = row_input_ptr + cols
-        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
-        g_ptrs = g_ptr + cols
-        output_ptrs = row_output_ptr + cols
-        x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
-        g = tl.load(g_ptrs, mask=mask, other=0.0)
-        rms_norm = x * norm_factor * g
-        tl.store(output_ptrs, rms_norm, mask=mask)
-
-
-# Main kernel with both blocked and non-blocked versions based on BLOCK_SIZE
 @triton.autotune(configs=get_autotune_config(), key=['n_rows', 'n_cols'], use_cuda_graph=True)
 @triton.jit
 def rms_kernel(output_ptr, input_ptr, g_ptr, input_row_stride, output_row_stride, n_rows, n_cols, epsilon,
                BLOCK_SIZE: tl.constexpr, USE_BLOCKED: tl.constexpr, NUM_PRGMS: tl.constexpr):
-    row_idx = tl.program_id(0)  # Each program instance handles one row
+    row_start = tl.program_id(0)  # Each program instance handles one row
     col_offsets = tl.arange(0, BLOCK_SIZE)
+    tl.assume(input_row_stride >= 0)
+    tl.assume(output_row_stride >= 0)
 
     if USE_BLOCKED:
-        # Blocked Approach: Accumulate sum of squares and normalize in chunks
-        sum_squares = accumulate_sum_squares(input_ptr, input_row_stride, n_cols, BLOCK_SIZE, row_idx)
-        mean_square = sum_squares / n_cols
-        norm_factor = tl.rsqrt(mean_square + epsilon)
 
-        # Apply normalization
-        apply_normalization(input_ptr, output_ptr, g_ptr, input_row_stride, output_row_stride, n_cols, norm_factor,
-                            BLOCK_SIZE, row_idx)
+        # Persistent loop for rows
+        for row_idx in tl.range(row_start, n_rows, NUM_PRGMS):
+            row_input_ptr = input_ptr + row_idx * input_row_stride
+            row_output_ptr = output_ptr + row_idx * output_row_stride
+
+            # Accumulate sum of squares
+            n_cols_blks = n_cols // BLOCK_SIZE
```

Comment:
Can this be hoisted out of the loop?

Reply:
put in the loop has slightly better perf